### PR TITLE
Rework & redesign settings menu

### PR DESCRIPTION
**Pull request by Reco1I**
In this PR I reworked and reorganized the entire settings menu UI following a section and categories pattern.

- [x] Depends on #349 

### Preview

image

### Sections
They're defined into the followings: 

* General: General usage and online/account options.
* Gameplay: Options that strictly affect user experience visually while gameplay or HUD options.
* Graphics: More technical than Gameplay section.
* Sound: Audio related options and volume controlls.
* Library: Beatmap library and storage related options, same as before.
* Input: Playfield size and synchronization options.
* Advanced: Same as before but some options where moved to the proper sections.

### Additional changes
There's some functionality changes that may need to be discussed (or maybe not):

* `Background quality` option was removed, this shouldn't be needed nowadays yet I don't think it was used at all.
* `Dither` option was removed, same reason as the above.
* `Complex animations` option was split in two: `Animate follow circle` and `Animate combo text`. Originally this option was covering those two features and I think that it's better to have those by separate.
* `Player name for offline` was removed and the value will be always "Guest", cuz more space.
* `Enable video` and `Enable storyboard` were added to `Gameplay` section.

### What's left

- [x] ~Apply the rework to the multiplayer settings menu (I guess is currently broken with this state)~

### CODE REVIEW

**Comment by Rian8337** [comment]
I'm not sure about removing any of the features mentioned. It definitely needs discussion.

**Comment by Reco1I** [comment]
> I'm not sure about removing any of the features mentioned. It definitely needs discussion.

Bet

File: src/com/reco1l/osu/ui/MainMenu.kt [hunk 1]
```diff
@@ -114,7 +113,7 @@ class MainMenu(val main: MainScene)
                     setColor(1f, 1f, 1f)
                     if (main.isOnExitAnim) return true
                     getGlobal().songService.isGaming = true
-                    getGlobal().mainActivity.runOnUiThread { SettingsMenu().show() }
+                    getGlobal().mainActivity.runOnUiThread { SettingsFragment().show() }
```
Comment (Rian8337) [inline src/com/reco1l/osu/ui/MainMenu.kt:116]:
```suggestion
                    mainThread { SettingsFragment().show() }
```

Maybe?

File: src/com/reco1l/osu/ui/SettingsFragment.kt [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+import ru.nsu.ccfit.zuev.skins.SkinManager
+import java.io.File
+
+
+enum class Section(@XmlRes val xml: Int) {
+
+    General(R.xml.settings_general),
+    Gameplay(R.xml.settings_gameplay),
+    Graphics(R.xml.settings_graphics),
+    Sounds(R.xml.settings_audio),
```
Comment (Rian8337) [inline src/com/reco1l/osu/ui/SettingsFragment.kt:61]:
```suggestion
    Audio(R.xml.settings_audio),
```

I think calling this `Audio` is more suitable.

File: src/com/reco1l/osu/ui/SettingsFragment.kt [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+        }
+
+
+        if (!Multiplayer.isMultiplayer) {
+            createSectionButton("General", R.drawable.grid_view_24px, Section.General)
+        }
+
+        createSectionButton("Gameplay", R.drawable.videogame_asset_24px, Section.Gameplay)
+        createSectionButton("Graphics", R.drawable.display_settings_24px, Section.Graphics)
+        createSectionButton("Sound", R.drawable.headphones_24px, Section.Sounds)
```
Comment (Rian8337) [inline src/com/reco1l/osu/ui/SettingsFragment.kt:149]:
```suggestion
        createSectionButton("Audio", R.drawable.headphones_24px, Section.Sounds)
```

Same here. I think `Audio` is better.

File: src/com/reco1l/osu/ui/SettingsFragment.kt [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+        }
+    }
+
+
+    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) = Unit
+
+
+    // For whatever reason this is restricted API when it wasn't in previous SDKs.
+    @SuppressLint("RestrictedApi")
+    override fun onBindPreferences() = when(section) {
```
Comment (Rian8337) [inline src/com/reco1l/osu/ui/SettingsFragment.kt:173]: This is one hell of a method. Maybe each section can be separated into its own method to make this method easier to parse mentally?

**Review by Rian8337 — COMMENTED** [review]
I have yet to test multiplayer, but I have some things to discuss with regards to the code and wording. I will not go deep into wording as even though this is a rework, there are many confusing wordings in the settings that fixing them warrants another PR, so as to not overcomplicate this PR more than it currently is.

File: res/xml/settings_advanced.xml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    </PreferenceCategory>
+
+    <PreferenceCategory android:title="Debug">
+
+        <CheckBoxPreference
+            android:defaultValue="true"
+            android:key="fps"
+            android:summary="@string/opt_fps_summary"
+            android:title="@string/opt_fps_title"
+            app:layout="@layout/settings_preference_checkbox_bottom" />
```
[the remark this answers is not in the document]
Comment (Reco1I) [reply res/xml/settings_advanced.xml:34]: Real-time PP, unstable rate and the rest of optional counters should not be dependent of this.

File: src/com/reco1l/osu/ui/SettingsFragment.kt (hunk 4 above)
Comment (Reco1I) [reply src/com/reco1l/osu/ui/SettingsFragment.kt:173]: That's overcomplicate things.

File: res/xml/settings_advanced.xml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Rian8337) [reply res/xml/settings_advanced.xml:34]: It should not, but that is not the point of my comment. Keeping this and leaving the dependency as is will produce misinformation.

File: src/com/reco1l/osu/ui/SettingsFragment.kt (hunk 4 above)
Comment (Rian8337) [reply src/com/reco1l/osu/ui/SettingsFragment.kt:173]: If we are going to add more functionality to settings that require explicit handling, the complexity of this method will grow at a faster rate.

Comment (Reco1I) [reply src/com/reco1l/osu/ui/SettingsFragment.kt:173]: No logic should be there.

> the complexity of this method will grow at a faster rate.

Before that we should take advantage of the listeners that `SharedPreferences` provides instead.

Comment (Rian8337) [reply src/com/reco1l/osu/ui/SettingsFragment.kt:173]: Sure, but that will add yet another complexity to this already large PR, so let's move that across for the time being.

**Comment by Rian8337** [comment]
I have tested multiplayer and it works properly.

**Comment by Rian8337** [comment]
I have moved the display FPS setting to "Gameplay" in line with other display settings.

Removing the setting's dependency should come in another PR.

**Review by Rian8337 — APPROVED** [review]
Big thank you for the work!

**Pull request merged**